### PR TITLE
[AC-434] Hide billing screen for reseller clients

### PR DESCRIPTION
**Pull request by r-tome**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
If an organization is managed by a Reseller, any Owner of that organization should only see billing information that is relevant to the organization. 

## Code changes
This PR is related to [this server PR](https://github.com/bitwarden/server/pull/2783).

- **apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html:** Hiding subscription information if User does not have permissions do manage billing
- **apps/web/src/app/organizations/billing/organization-subscription-cloud.component.ts:** Getting subscription info even if user cannot manage billing
- **apps/web/src/app/organizations/billing/organization-subscription-selfhost.component.html:** Changed check to variable `canAccessBilling`
- **libs/common/src/enums/providerType.ts:** Enum containing Provider types
- **libs/common/src/models/data/organization.data.ts:** Added `ProviderType` property
- **libs/common/src/models/domain/organization.ts:** Added `ProviderType` property 
- **libs/common/src/models/response/profile-organization.response.ts:** Added `ProviderType` property

## Screenshots
image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/common/src/models/domain/organization.ts [hunk 1]
```diff
@@ -196,8 +199,20 @@ export class Organization {
...
+        this.providerType == ProviderType.Reseller)
+    );
+  }
+
   get canManageBilling() {
-    return this.isOwner && (this.isProviderUser || !this.hasProvider);
+    return (
+      this.isOwner &&
+      ((this.isProviderUser && this.providerType == ProviderType.Msp) || !this.hasProvider)
+    );
```
Comment (eliykat) [inline libs/common/src/models/domain/organization.ts:215]: I don't understand this boolean logic - it's pretty hard to parse. I also think the names are unclear.

I suggest breaking it down into individual permissions to better express what's going on here:

* `canEditSubscription` - returns true if
  * `(isOwner && !hasProvider)` - you're an owner of an org with no provider, OR
  * `(isProviderUser && hasProvider)` - you're a providerUser managing that org (implicitly an MSP)
  * this means that you can't edit the sub if you're a Provider client (MSP or Reseller)
* `canViewSubscription` - returns true if 
  * `(isOwner && !hasProvider)` - you're an owner of an org with no provider, OR
  *  `(isOwner && hasProvider && providerType == Reseller)` - you're a Reseller client
* `canEditPaymentMethods` - same as canEditSub
* `canViewBillingHistory` - same as canEditSub

Then the content follows these rules:
* if you can view the sub, you can see the basic subscription details on the Subscription page. If not, you get the billing obfuscation image (the `app-org-subscription-hidden` component)
* if you can view & edit the sub, you also get the rest of the subscription details
* if you can edit payment methods, you can access that tab
* if you can view billing history, you can access that tab

I think that works but please double check me and make sure it works.

There are 2 secondary things I'm trying to do here:
* use more granular permissions (not just "billing" but different aspects of billing)
* use clear create/edit/view/delete language rather than "manage" which has always been unclear to me

File: apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html
Comment (eliykat) [inline apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html:1]: You're still showing the SM Beta opt-in, is that correct?

**Review by eliykat — CHANGES_REQUESTED** [review]

File: apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html
Comment (r-tome) [reply apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html:1]: It is not, I've verified with Sukhleen. Its now hidden.

File: apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html [hunk 2]
```diff
@@ -68,27 +68,29 @@ <h1>
...
       </div>
+      <ng-container *ngIf="userOrg?.providerId != null">
+        <div class="col-sm">
+          <dl>
+            <dt>{{ "provider" | i18n }}</dt>
+            <dd>{{ "yourProviderIs" | i18n: userOrg.providerName }}</dd>
+          </dl>
+        </div>
+      </ng-container>
     </ng-container>
```
Comment (eliykat) [inline apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html:94]: Take a look at line 86/87 above (the nested `ng-container`). It displays some provider information, but now that's going to be hidden if `!canEditSubscription`, which is going to be anyone with a provider. So that info will never be visible.

But - that's actually a problem at the moment. If `!manageBilling` then the `sub` never loads and this section is not shown. Only the provider (MSP) would ever see it, and they already know who the provider is 😂  Can you check if we can just delete it? Just so we don't have dead templating hanging around.

File: apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html
Comment (eliykat) [inline apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html:1]: Please stick an `*ngIf="userOrg.canEditSubscription"` on line 94/96 (the `ng-container` wrapping the buttons). It looks like the `showChangePlan` and `ShowChangePlanButton` getters will hide them in this case anyway, but it's easier and safer to make this explicit.

**Review by eliykat — CHANGES_REQUESTED** [review]

File: libs/common/src/models/domain/organization.ts (hunk 1 above)
Comment (eliykat) [reply libs/common/src/models/domain/organization.ts]: I really like this change! Boolean logic gets confusing really quick, I like using ternary expressions to simplify it where possible.

File: apps/web/src/app/organizations/billing/organization-subscription-selfhost.component.ts [hunk 3]
```diff
@@ -101,7 +101,7 @@ export class OrganizationSubscriptionSelfhostComponent implements OnInit, OnDest
     }
     this.loading = true;
     this.userOrg = this.organizationService.get(this.organizationId);
-    if (this.userOrg.canManageBilling) {
+    if (this.userOrg.canEditSubscription) {
```
[the remark this answers is not in the document]
Comment (r-tome) [reply apps/web/src/app/organizations/billing/organization-subscription-selfhost.component.ts:104]: Correct, fixed to also use `canView`.

File: apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html (hunk 2 above)
Comment (r-tome) [reply apps/web/src/app/organizations/billing/organization-subscription-cloud.component.html:94]: Ooh nicely spotted! Its not in the Figma so I think we want to delete it. I've asked Sukhleen

File: libs/common/src/models/domain/organization.ts [hunk 4]
```diff
@@ -196,8 +199,22 @@ export class Organization {
     return this.canManagePolicies;
   }
 
-  get canManageBilling() {
-    return this.isOwner && (this.isProviderUser || !this.hasProvider);
+  get canViewSubscription() {
+    return this.hasProvider && this.providerType === ProviderType.Msp
+      ? this.isProviderUser
+      : this.isOwner;
+  }
```
Comment (eliykat) [inline libs/common/src/models/domain/organization.ts:206]: This is what I meant in my previous comment: Edit implies (gives permissions to) View

```suggestion
  get canViewSubscription() {
    if (this.canEditSubscription) {
      return true;
    }
    
    return this.hasProvider && this.providerType === ProviderType.Msp
      ? this.isProviderUser
      : this.isOwner;
  }
```

I think this is cleaner than the other way around, where Edit might fail because they don't have View. Instead, assume that if they can Edit then they can View.

File: libs/common/src/models/domain/organization.ts [hunk 5]
```diff
@@ -196,8 +199,22 @@ export class Organization {
...
+
+  get canEditSubscription() {
+    return this.canViewSubscription && (this.hasProvider ? this.isProviderUser : this.isOwner);
```
Comment (eliykat) [inline libs/common/src/models/domain/organization.ts:209]: Then revert this recent change:

```suggestion
    return this.hasProvider ? this.isProviderUser : this.isOwner;
```

**Review by eliykat — CHANGES_REQUESTED** [review]
Most recent changes look good, only outstanding items are:
1. my feedback below (to clarify my previous suggestion)
2. removing unauthorised data from the subscription model.

File: libs/common/src/models/domain/organization.ts (hunk 4 above)
Comment (r-tome) [reply libs/common/src/models/domain/organization.ts]: That makes perfect sense!

File: apps/web/src/app/billing/organizations/organization-subscription-cloud.component.html [hunk 6]
```diff
@@ -60,8 +60,7 @@ <h1>
             {{
               nextInvoice
                 ? (nextInvoice.date | date : "mediumDate") +
-                  ", " +
-                  (nextInvoice.amount | currency : "$")
+                  (userOrg.canEditSubscription ? ", " + (nextInvoice.amount | currency : "$") : "")
```
Comment (eliykat) [inline apps/web/src/app/billing/organizations/organization-subscription-cloud.component.html:63]: Andreas made some changes to this page here: https://github.com/bitwarden/clients/pull/4897 We should make sure not to accidentally overwrite his work.

It looks like your feature branch is out of date, please merge master -> feature, then feature -> this branch and resolve merge conflicts. Then I can approve :)

**Review by eliykat — CHANGES_REQUESTED** [review]

File: apps/web/src/app/billing/organizations/organization-subscription-cloud.component.html (hunk 6 above)
Comment (r-tome) [reply apps/web/src/app/billing/organizations/organization-subscription-cloud.component.html]: The amount is gone which simplifies this a lot. Updated!

**Review by eliykat — APPROVED** [review]
👏
